### ISSUE TITLE
[Proposal] Add a low-level `Communication` module

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#17657]
This issue proposes addition of a low-level communication module which initially
allows point-to-point raw data copies.

### Module name: `Communication`

- It captures the main purpose of the module
- It can contain collectives in the future either as top-level functions or in a
  submodule
- Alternatives I thought about:
  - Comm
    - I like this, but the user can also do `import Communication as Comm`
    - Consistent with CommDiagnostics module, though it is not a stable
      interface
  - Multilocale
    - Too general, doesn't tell much
  - LowLevelCommunication or LowLevelComm
    - Not sure if we need "LowLevel" in the name
  - SPMD
    - Not necessarily limited to SPMD-like idioms

### Interface

I propose we only have 1 function in the first step:

```chapel
proc copy(ref dst: ?t, const ref src: t, numElems: int, useCache=true)
```

which copies `numElems*numBytes(t)` bytes from the `src` address to `dst`,
regardless of whether `src`/`dst` being local/remote.

#### Name

- We don't want to have `get`/`put` as they only differ w.r.t. location
  of source/destination. Let's not make the user worry about it.
- On similar things we use "copy":
  - `c_memcpy`
  - we call the compiler-driven aggregation "Automatic Copy Aggregation"
  - We have the `UnorderedCopy` module (which we might incorporate into this
                                        module in the future. See below)
  - Arkouda's aggregators have a `copy` method

#### Arguments

- I think we should have `numElems` not `numBytes`
- `dst`/`src` is consistent with C's `memcpy` and sufficiently clear to me.
  - I wouldn't oppose to `lhs`/`rhs`, too. But they are too specific to
    assignments, which may not be how an MPI-minded programmer think about this
    operation.
- `useCache`: I think we can do some overloading tricks to check for
  `--cache-remote`, as we don't want `useCache=true` with `--no-cache-remote`.
  Similar to controlling the caching behavior, we may need more kinds of `copy`s
  in the future. See below.
- The contexts in which the need for this function came up has been about
  primitive types, so far. We may want to extend to POD/non-POD types down the
  road but current interface shouldn't block that from happening I think. See
  below for more discussion about data types.

---

Some of the finer details are below. There are also some questions about the
future additions to this module. I don't think we have to answer any of those at
the moment. But it might help to have the bigger context handy when making
decisions

### How should we manage different kinds of copies?

In the near term we want to control:

  - cached/uncached

In the longer term we might want to have

  - unordered
  - nonblocking
  - strided
  - aggregated (maybe? but probably it is going to be a separate thing)

I think we should expose these things via arguments rather than implementing
different functions. Though I have some uncertainties:

- nonblocking version might want to return a status object of sorts. Changing
  the return type of the function based on an argument may not be a good idea.
  The alternative would be adding something like `copyNB` or `nonBlockingCopy`
  or something else. Another alternative that I can think of is to have
  `Communication.NonBlocking` submodule.

  Same question about nonblocking-ness can come up with collectives, when we add
  them, too.

- strided version may require more than one argument, which may not be too
  difficult to handle but it is something to consider.

- aggregated copies are probably going to be handled via the aggregator objects.
  However, I mused about passing that object to the `copy` here as an
  alternative instead of having that object have its own user-facing `copy`
  method. See https://github.com/chapel-lang/chapel/issues/16963 for user-facing
  aggregation discussion.

#### POD types

- I think we can extend to POD types. But with serialization and copy/move
  semantics for user-defined records, things get murky.

  For non-primitive types I see the following alternative paths for this module:

  - Follow the semantics of corresponding scalar/array assignment for the given
    type.

    - Pro: consistent with language rules, widest coverage for this module
    - Con: what if that assignment is not in bulk, or require non-O(1) local
           operations?  Doesn't it go against the nature of this module?

  - Follow the semantics of corresponding scalar/array assignment for the given
    type *if* that assignment can be handled with bulk transfer

    - Pro: more guarantee towards O(1) operation/communication complexity?
    - Con: understanding whether something is bulk transferrable may not be easy
           for the user

  - Limit only to primitive types (what I propose as the first step here)

    - Pro: Safest to say "one call to `copy` is negligable local ops + 1 comm"
    - Con: too limiting?

#### Serializable types

- I think we can do this, too. But we may want to think about bulk
  serialization/deserialization first.

  See https://github.com/chapel-lang/chapel/issues/15675, https://github.com/chapel-lang/chapel/issues/15676

#### Class types

- Currently, we only allow bulk copy of unmanaged instances (where we do a bulk
  pointer copy), I think we can also extend this interface to those cases. I am
  not sure how important that is. The reason we have added bulk copy of
  unmanaged class arrays was to have faster distributed array initialization.

  (Bulk class transfer is still very limited, because we only wanted it to work
  for array creation in the past, and we haven't come across a case to make it
  more general, yet. Though, I think it should be easy to do. See https://github.com/chapel-lang/chapel/pull/15049)

---
  
### Summary of future work

- I would like this module to subsume the `UnorderedCopy` package module. I
  imagine we can do that by adding an argument to `copy`
- We should also think about moving `UnorderedAtomics` somewhere in this module
  or to a sibling/child of sorts. Maybe something like
  `Communication.UnorderedAtomics` ?
- We should think about non-blocking versions of the operations.
- We want to have some basic collectives for reduction, broadcast and barriers
  - We'll explore the idea of task teams/groups as a high-level library concept
- User-facing aggregators should at least be consistent with this interface. But
  IMO they can be addition to this module as a top-level concept or a submodule.
  See https://github.com/chapel-lang/chapel/issues/16963

**Comment by e-kayrakli** [comment]
@mppf -- You have mentioned that this module is related to split-init and copy elision across locales. Can you elaborate a bit more on that?

**Comment by mppf** [comment]
@e-kayrakli it's more that it's related to the issues #15675, #15676, and #15808. Part of the reason we don't have split-init or copy-elision across locales today is that we didn't have a way to allow a record to adjust to such movement. A record might contain a C pointer that is always local, for example. Anyway, even if we never allow split-init or copy-elision to work directly with `on` statements, we should hopefully have a way a user can opt in to that. Which might end up being a call to this module.

**Comment by gbtitus** [comment]
I would argue strongly against `lhs` and `rhs`. They force the user to figure out what thing's left-hand side we're talking about. `dst` and `src` are much better.

**Comment by mppf** [comment]
> proc copy(ref dst: ?t, const ref src: t, numElems: int, useCache=true)

I think it'd be fine to have `proc copy(ref dst: ?t, const ref src: t, useCache=true)`. But when it comes to copying multiple elements, I go "eww" when we pass the pointer to multiple elements with a reference. The reason being that in typical use of `ref` in Chapel, the `ref` is pointing to only a single element. The implementation has to turn the `ref` into a pointer in order to even access elements beyond the one pointed to. I understand that there are other places that we do this (especially for interoperability, but I don't like it there either). I also understand that the language makes it more convenient to get a reference to something (since it happens implicitly) than to get a "pointer".

But, we can't use `c_ptr` for these arguments because they are intended to be on different locales in typical usage, and passing locales separately seems like unnecessary work. At the same time, we can't use `_ddata` for it because that's not a user facing type.

So, I think that we need to have some kind of low-level pointer (that is typically wide in multilocale settings). This is requested/discussed in #16797, #12490, #8680. I think that `copy` should accept the low-level pointer type for the `dst` and `src` arguments.

> useCache: I think we can do some overloading tricks to check for
> --cache-remote, as we don't want useCache=true with --no-cache-remote.
> Similar to controlling the caching behavior, we may need more kinds of copys
> in the future. See below.

I think that the `useCache` argument should just be ignored when we have `--no-cache-remote`. I think it'd be overly fussy to make it a compile-time or run-time error to use `--no-cache-remote` and `useCache=true`.

> strided version may require more than one argument, which may not be too difficult to handle but it is something to consider.

It seems to me you would like to have, instead of `numElems: int`, a range or domain (or tuple of ranges), to handle the strided case. I think it would be interesting to consider if the interface would still be clean enough with a range for the non-strided cases.

> POD types

Did you mean for this section to be called "Record types"?

> Currently, we only allow bulk copy of unmanaged instances (where we do a bulk pointer copy), 

`unmanaged` class types are POD, so I think these should be handled just like any other POD data. A non-nilable unmanaged class type can't be default inited, but that could also be true for a record that is trivially copy-initializeable/assignable. (I tend to use "POD" as shorthand for trivially copy-initializable/assignable but the latter is probably a better term).

### module organization

> Module name: Communication

I have a counter-proposal that I'd like to discuss (although I don't feel very strongly about this). The
[Memory.Initialization](https://chapel-lang.org/docs/modules/standard/Memory/Initialization.html) module already has moveInitialize and moveInitializeArrayElements. I think it'd be reasonable to extend these to work in multilocale settings. (However I think they should accept pointers rather than just arrays).

When it comes to nonblocking copies or copies with or without caching, having the functionality in a Communication module sounds more reasonable. But I'm uncertain where to draw the boundary between the modules. It seems strange to me that you'd use Memory.Initialization for local initialization and Communication for remote initialization, say.

### bulk transfer

Re bulk transfer, I think that broad support for bulk transfer is really important. I think we can figure out a variation on the existing bulk transfer mechanism that includes serialize/deserialize. E.g, to transfer `n` elements with serialize/deserialize:
 * create a temporary `src `buffer consisting of a chunk of up to `n` of the type that `serialize` produces
 * fill in the temporary buffer by calling `serialize` on the elements
 * "on" the destination locale:
    * create a temporary `dst` buffer consisting of `n` of the type that `serialize` produces
    * copy the bits of the serialized data across the network from `src` to `dst`
    * initialize the elements in the destination array with the result of `deserialize` applied to the `dst` buffer elements
    * deallocate the `dst` buffer (but don't deinitialize the elements)
 * deinitialize the `src` buffer elements (or, maybe this is `deserialize`'s job?)
 * deallocate the `src` buffer
 * repeat for any remaining chunks

That's a sketch. For copy vs move, the difference is that "move" will  "consume" the original element while for "copy" the original element must be preserved. Anyway I think that minor variations on this recipe would make a big difference for performance and so that's worth investigating. It would be interesting to consider comm/compute overlap between the serialize/deserialize process and using the network, among other things. In terms of interface design, I think one key bit of information here is that the serialize/deserialize process will necessarily involve an `on` statement when doing the bulk transfer, while if serialize/deserialize are not necessary, it can just be a PUT/GET (at least if the elements are contiguous).

**Comment by bradcray** [comment]
Capturing some conversation around this module from the performance meeting yesterday:

* Q: What would motivate a user to use the `copy()` routine rather than `=`?  A: the ability to pass flags like those related to the cache.
* For various reasons, we discussed starting with `get()` and `put()` calls as stepping stones towards an eventual `copy()`, where I'd characterize the discussion as follows:
  * they seem more obviously like communication calls; less "why am I not just using `=` again?" reactions
  * they avoid any need to determine whether one, both, or neither argument is remote by definition
  * they seem more natural to take `c_ptr`, `c_ptr`, `locale` arguments, dodging the "multiple elements" challenges noted above
  * they would be sufficient for us to stop using communication `__primitives()` in most user codes and benchmarks today
* we discussed the challenge of (generally) not being able to assume that array elements are laid out in any particular way, and the benefit of having more introspection capabilities along the lines of "is contiguous", "is local", etc. for reasoning about array memory
* we discussed the relationship between multi-element bulk transfers and slicing—slices are often a natural way to express sub-arrays (as well as the queries of the previous bullet), but are currently too expensive
* we also discussed that some cases where we use communication `__primitives()` may be able to be fixed fairly easily by having the bulk transfer code check whether the LHS is remote before using its current/default implementation

**Comment by e-kayrakli** [comment]
Thanks Brad for the notes. I'll add that we also had a conversation around `useCache` and similar flags that control the cache behavior.

- @ronawho mentioned that `useCache=false` shouldn't necessarily mean "don't touch cache at all" but instead we may need to fence the cache globally to make sure that we are not touching memory locations that have pending cached operations. (I believe this is what we do today with transfers that are larger than some threshold)
- There's been some discussions around whether the low-level nature of this module implies a "don't touch cache at all" behavior, and whether we can provide a way in which that behavior can be provided.
  - A related idea was to add another flag like `useCache` that controls whether we should fence or not
    - (... but if the function(s) is called in a tight loop where every time we fence, we would be paying too much for unnecessary fences)
  - Another option is to expose the fence in a separate function.
  - One option is to have some sort of a "cacheless region" where cache is implicitly fenced on entry (and exit?). This can be either a function pair `begin*()`/`end*()` or a context manager.

**Comment by mppf** [comment]
> @ronawho mentioned that useCache=false shouldn't necessarily mean "don't touch cache at all" but instead we may need to fence the cache globally to make sure that we are not touching memory locations that have pending cached operations. (I believe this is what we do today with transfers that are larger than some threshold)

Actually we just invalidate the regions being communicated with the GET/PUT that is too big to run through the cache. E.g.

https://github.com/chapel-lang/chapel/blob/c95d6a9e204e638158d0b87b6f66b98fafac9c18/runtime/src/chpl-cache.c#L3746-L3748

Using `get` and `put` also more or less sidestep the issues I was raising about whether copy/move initializers or serialize/deserialize functions are called. (In particular I wouldn't expect put/get to serialize, deserialize, or call a copy initializer).

**Comment by e-kayrakli** [comment]
The direction we're going seems to be adding a `Communication` module with `put`/`get` standalone functions in it as the first step.  


### Summary of what I am proposing:

```chapel
module Communication {
  proc get(dest: c_ptr, src: c_ptr, target: int, size: int) 
  proc put(dest: c_ptr, src: c_ptr, target: int, size: int)
}
```

### Four arguments and their types:

#### Two addresses
It looks pretty clear that they need to be `c_ptr`s.

#### One argument for some sort of locale information
I propose we have an `int` argument for locale ID. We can think of adding an overload that takes a locale type, as well. Currently, all our modules' uses of `chpl_comm_get` and `chpl_comm_put` primitives have locale ID readily available. If we have only a `locale`-argument version, we'll need to change those call sites.

#### One argument for comm size
I propose we make this argument be number of bytes rather than number of elements, given the low level nature of these functions. However, I am on the fence for this one -- we can grab the type from the `c_ptr` argument, and ask the user to give us number of elements instead. I don't have a strong preference here.

### Argument names

I propose we go with `dest`, `src`, `target`, `size`:

- `dest` and `src`, I grabbed from `c_memcpy`. I don't think this module is that stable, we can also consider spelling the words out completely, or using `dst` instead of `dest`
- `target` is something we use already in `targetLocales`. We also call where an `on` executes its `target`, for example. Note that this argument means where `src` is on a `get`, and where `dest` is on a `put`.
- `size` is what we use for our arrays, data strucutures and arrays.


### Examples from other languages/libraries:

#### MPI:

```c
int MPI_Get(void *origin_addr, int origin_count, MPI_Datatype
            origin_datatype, int target_rank, MPI_Aint target_disp,
            int target_count, MPI_Datatype target_datatype, MPI_Win
            win)

int MPI_Put(const void *origin_addr, int origin_count, MPI_Datatype
            origin_datatype, int target_rank, MPI_Aint target_disp,
            int target_count, MPI_Datatype target_datatype, MPI_Win
            win)
```

#### SHMEM

```c
void shmem_get(TYPE *dest, const TYPE *source, size_t nelems, int pe);
void shmem_put(TYPE *dest, const TYPE *source, size_t nelems, int pe);
```

#### UPC

```c
void upc_memget(void * restrict dst, shared const void * restrict src, size_t n);
void upc_memput(shared void * restrict dst, const void * restrict src, size_t n);
```

#### UPC++

```c++
template < typename T , typename Cx=/*...*/)
RType rput (T const * src , global_ptr <T > dest , size_t count ,
            Cx &&completions=operation_cx::as_future());

template < typename T , typename Cx=/*...*/>
RType rget (global_ptr < const T > src ,
            Cx &&completions=operation_cx::as_future());
```

**Comment by vasslitvinov** [comment]
I suggest renaming `target` to `srcLoc` / `dstLoc` / `destLoc` (to match 'dst' vs. 'dest'). "Target" is rather vague in this context.

Given that these functions are close in meaning to `c_memcpy`, I propose that the naming and the meaning of the arguments (excluding the locale arg) match among these three functions.

Speaking of which, the `n` argument of `c_memcpy` should have a better name. (Unless the name `n` was the consensus at CTypes' module review.)

**Comment by ronawho** [comment]
I also don't like `target` as the argument name. `loc` /`srcLoc`/`dstLoc`/`localeId` would be fine with me. I'm fine keeping it an `int` though since`locale.id` returns an `int`.

I buy Michael's argument that the types should be `c_void_ptr` and also agree it'd be good to have a user-facing way to get from a potentially wide chapel pointer to the c_ptr.

For Arkouda aggregation primitives I have something like:

```chpl
  inline proc getAddr(const ref p): c_ptr(p.type) {
    // TODO can this use c_ptrTo?
    return __primitive("_wide_get_addr", p): c_ptr(p.type);
  }
```

Where I never checked if just using `c_ptrto` would do the right thing.

I also agree that symmetry with memcpy would be nice, but I'm open to changing the names so long as we're consistent across. Could you take a look at the gasnet API too and add that to your list of other libraries?

**Comment by vasslitvinov** [comment]
> A related issue is that if you have a wide pointer it is currently challenging to split apart the components of the wide pointer

I suggest addressing this issue by creating overload(s) accepting wide or narrow pointer(s) directly. Perhaps via generic [const] ref argument(s)? Having the user split them manually is non-chapeltastic.

Echoing Michael's comments earlier, overloads accepting Chapel types should report an error for non-POD types, at least for now. For POD types, there is similarity with th eMemory.Initialization module.

> I also agree that symmetry with memcpy would be nice, but I'm open to changing the names so long as we're consistent across.

This is what I meant to say, too.

**Comment by mppf** [comment]
> I agree with all here. I was thinking that `c_ptrTo` could do that, but apparently it calls an extern, precluding passing wide arguments to it. Hmm.. is it appealing to keep the destination and source arguments generic, and grab the pointer via primitives inside the function?

I don't think it should call an extern function. It should use a primitive.

**Comment by e-kayrakli** [comment]
"Should"? maybe, yes. But currently it ends up calling `extern proc c_pointer_return`, which I think we can do without.

**Comment by bradcray** [comment]
I also had a hesitant reaction to the argument name `target`, though my thought was to switch to `targetLoc` or `remoteLoc` or `targetLocID` or `remoteLocID` or `locID`.  I'm also not opposed to putting src/d[e]st into those names which I think would have the benefit of slightly more clarity at the cost of asymmetry.  If these take ints, I suspect we'll want to use the argument name, so may want to keep it short/sweet.

I'd also be inclined to change the size argument to indicate the units like `numBytes` or `sizeInBytes` or `numElts`.  Again, this seems like an argument a programmer interested in readability would want to type, esp. if we end up with two integer arguments.

I'm pretty open to having these take integers rather than locales on the argument that they're pretty low-level / systems-y.  If we had versions that took generic/`ref` Chapel types and computed the pointer for you, I'd be more inclined to have those overloads take locales since they're avoiding C-/system-level information.

**Comment by bradcray** [comment]
@e-kayrakli / @ronawho : Do you think this can be closed with #20309 merged?  The PR referred to it but didn't close it, so I'm guessing it's intentional that it's still open.  Were there outstanding action items on this issue that are keeping it open?  Could we fork those off to their own issues since it seems like the high-order bit in the OP was achieved?

**Comment by e-kayrakli** [comment]
I am OK with closing it. I kept it open in the past because we may want to consider a general `copy` function as proposed in the OP. But I am personally happy with where we are today and probably consider more incremental improvements to the module (avoid cache, consider sublocales, non-blocking comm etc) before thinking about a `copy` interface.
